### DISCUSSION TITLE
Updating same object from two different tasks in async

### DISCUSSION THREAD

**Discussion by pankajkumar229** [sqlalchemy/sqlalchemy#9201] [Usage Questions] [answered]
I am encountering an exception when trying to add an object and commit it using async sqlalchemy.

The exception looks like this:
sqlalchemy.exc.InvalidRequestError: Object '<User at 0x7ff0004bb290>' is already attached to session '6' (this is '7')

Each request handler gets its own scoped session with currenttask scope. IF the same user does two API calls concurrently, this is bound to happen. DB itself is not a problem when doing concurrent updates (the other session is only reading). What am I doing wrong here? Please see the full stack below.

```
Traceback (most recent call last):
  File "/product/labgrader/nbgrader/auth.py", line 114, in wrapper
    retVal = await func(self, *args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/product/labgrader/nbgrader/apis/dbapi.py", line 1276, in put
    return await self.put_as_student(user, course_id, assignment_id)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/product/labgrader/nbgrader/apis/dbapi.py", line 1099, in put_as_student
    await self.gradebook_async.upsert_student_information_to_assignment(assignment_id, user.id, {}, **assignment_mapping)
  File "/product/labgrader/nbgrader/apiasync.py", line 698, in upsert_student_information_to_assignment
    assignment_student_association.user = student
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/sqlalchemy/orm/attributes.py", line 460, in __set__
    self.impl.set(
  File "/usr/local/lib/python3.11/site-packages/sqlalchemy/orm/attributes.py", line 1269, in set
    value = self.fire_replace_event(state, dict_, value, old, initiator)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/sqlalchemy/orm/attributes.py", line 1295, in fire_replace_event
    value = fn(
            ^^^
  File "/usr/local/lib/python3.11/site-packages/sqlalchemy/orm/unitofwork.py", line 130, in set_
    sess._save_or_update_state(newvalue_state)
  File "/usr/local/lib/python3.11/site-packages/sqlalchemy/orm/session.py", line 2667, in _save_or_update_state
    self._save_or_update_impl(state)
  File "/usr/local/lib/python3.11/site-packages/sqlalchemy/orm/session.py", line 3284, in _save_or_update_impl
    self._update_impl(state)
  File "/usr/local/lib/python3.11/site-packages/sqlalchemy/orm/session.py", line 3267, in _update_impl
    to_attach = self._before_attach(state, obj)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/sqlalchemy/orm/session.py", line 3361, in _before_attach
    raise sa_exc.InvalidRequestError(
sqlalchemy.exc.InvalidRequestError: Object '<User at 0x7ff0004bb290>' is already attached to session '6' (this is '7')
```

**Comment by zzzeek** [comment, marked as answer]
it seems likely that however you are using a "scoped session" under async, which is an intricate technique I'd not recommend, is not correctly keeping sessions isolated.  

the normal SQLAlchemy scoped_session uses a threading.local() and will not work at all with async.   for async, there is an [async_scoped_session](https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#asyncio-scoped-session)  which will use asyncio.current_task as the "scopefunc", but I don't have experience using "current task" as such a function and it's not known if this is safe to use for any particular async web framework.

short answer I'd definitely absolutely **not** use ``scoped_session()`` with async, and if using ``async_scoped_session()``, I'd be extremely careful and still prefer not to use it.  it's best if you create an async_session at the start of a request and pass it to requests explicitly.  if using a web framework where such an option is not available, seek help from the maintainers of that framework on the best way to have resources that are safely scoped to a particular request and use that technique.

**Reply by pankajkumar229** [reply]
Hi Mike

I am trying to use async_scoped_session with current_task as the scope function. I would have to rewrite a lot to not use it :( 

IF there is any way to make it safe, please let me know.

Thanks,

**Reply by zzzeek** [reply]
it depends on the web framework in use and the specifics of the code, including how objects are handled.  in particular if you are calling session.add(user), or associating "user" with an object that's in that session as is what you seem to be doing and getting that error, that implies this "user" object came from a different Session.     within the request, all data should be loaded from one and only one Session linked to that request.

**Reply by pankajkumar229** [reply]
Let me check by printing the task ids but I am pretty sure the user object was retrieved in the same task.

**Reply by pankajkumar229** [reply]
My apologies for doubting the framework first. IT was our code that was wrong. Although both objects were called from the same task, we are using two async_scoped_session per task (they have different isolation levels). The objects were mixed and SQLAlchemy gave the exception. Thanks much for your help.
